core/src/main/java/io/substrait/expression/FunctionArg.java

```diff
@@ -0,0 +1,101 @@
+package io.substrait.expression;
+
+import io.substrait.expression.proto.ProtoExpressionConverter;
+import io.substrait.function.SimpleExtension;
+import io.substrait.proto.FunctionArgument;
+import io.substrait.type.Type;
+import io.substrait.type.TypeVisitor;
+import io.substrait.type.proto.FromProto;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.stream.IntStream;
+
+public interface FunctionArg {
+
+  <R, E extends Throwable> R acceptFuncArgVis(FuncArgVisitor<R, E> fnArgVisitor) throws E;
+
+  interface FuncArgVisitor<R, E extends Throwable> {
+    R visitExpr(Expression e) throws E;
+
+    R visitType(Type t) throws E;
+
+    R visitEnumArg(EnumArg e) throws E;
+  }
+
+  static <R, ER, TR, E extends Throwable> FuncArgVisitor<R, E> createFuncArgVisitor(
+      TypeVisitor<TR, E> typeVisitor,
+      Function<TR, R> convertTyp,
+      ExpressionVisitor<ER, E> expressionVisitor,
+      Function<ER, R> convertExpr,
+      Function<EnumArg, R> convertEnum) {
+    return new FuncArgVisitor<R, E>() {
+
+      public R visitExpr(Expression e) throws E {
+        return convertExpr.apply(e.accept(expressionVisitor));
+      }
+
+      @Override
+      public R visitType(Type t) throws E {
+        return convertTyp.apply(t.accept(typeVisitor));
+      }
+
+      @Override
+      public R visitEnumArg(EnumArg e) throws E {
+        return convertEnum.apply(e);
+      }
+    };
+  }
+
+  class ProtoFrom {
+    private final ProtoExpressionConverter exprBuilder;
+
+    public ProtoFrom(ProtoExpressionConverter exprBuilder) {
+      this.exprBuilder = exprBuilder;
+    }
+
+    public FunctionArg convert(
+        SimpleExtension.Function funcDef, int argIdx, FunctionArgument fArg) {
+      return switch (fArg.getArgTypeCase()) {
+        case TYPE -> FromProto.from(fArg.getType());
+        case VALUE -> exprBuilder.from(fArg.getValue());
+        case ENUM -> convert(funcDef, argIdx, fArg.getEnum());
+        default -> throw new UnsupportedOperationException(
+            String.format("Unable to convert FunctionArgument %s.", fArg));
+      };
+    }
+
+    /**
+     * Account for optional arguments before argument position. For example for `(opt, req, opt,
+     * req)`: - for `(v1)` `v1` can be `arg0 or arg1` - for `(v1, v2)` `v2` can be `arg1, arg2,
+     * arg3`
+     *
+     * <p>TODO: move this to `SimpleExtension.Function`, or at least an inner function in
+     * `possiblePositions`
+     *
+     * @param args
+     * @param argIdx
+     * @return
+     */
+    private int possiblePositions(List<SimpleExtension.Argument> args, int argIdx) {
```

Comment:
I think there is a misunderstanding here of the nature of optional enum arguments. They must always be declared when using a function. (There should be no ambiguity on the consumption of the plan as to what each argument means.) The optionality is to declare an enum that has no value.

For example, if you have a function that was `foo(Optional Enum, Required Enum, Value, Value, Type)` you always have to pass the optional enum. You can either provide a value or no value. If you provide no value, the default is used. The argument must always be passed, however.

Part of the misunderstanding is that isthmus currently fails to behave correctly I think...

Reply:
Thanks, this helps. Yes, I took optional to me possibly not present in the invocation. Will fix.